src/sc/sidechainTxsCommitmentBuilder.cpp

```diff
@@ -0,0 +1,203 @@
+    return field;
+#endif
+#if IMPLEMENTATION == 2
+    std::vector<unsigned char> mediumTerm(SC_FIELD_SIZE, '\0');
+    mediumTerm.insert(mediumTerm.begin(), ccoutHash.begin(), ccoutHash.end());
+    mediumTerm.insert(mediumTerm.begin()+ccoutHash.size()/sizeof(unsigned char), txHash.begin(), txHash.end());
+
+    unsigned char* outPosPtr = static_cast<unsigned char*>(static_cast<void*>(&outPos));
+    std::vector<unsigned char> tmp(outPosPtr, outPosPtr + sizeof(unsigned int));
+    std::move(std::begin(tmp), std::end(tmp), mediumTerm.begin()+(ccoutHash.size()+txHash.size())/sizeof(unsigned char));
+    mediumTerm.resize(SC_FIELD_SIZE, '\0');
+
+    LogPrintf("%s():%d - Data about to be mapped to field point [%s]\n",
+        __func__, __LINE__, HexStr(mediumTerm));
+
+    //generate field element. MISSING null-check
+    field_t* field = zendoo_deserialize_field(mediumTerm.data());
+    return field;
+#endif
+#if IMPLEMENTATION == 3
+    std::string toHash = std::string(ccoutHash.begin(), ccoutHash.end()) +
+            std::string(txHash.begin(), txHash.end()) +
+            std::string(static_cast<char*>(static_cast<void*>(&outPos)), sizeof(unsigned int));
+
+    unsigned char mediumTerm[SC_FIELD_SIZE] = {};
+    memcpy(mediumTerm, toHash.append(SC_FIELD_SIZE - toHash.size(), '\0').c_str(), SC_FIELD_SIZE);
+
+    LogPrintf("%s():%d - Data about to be mapped to field point [%s]\n",
+        __func__, __LINE__, HexStr(mediumTerm));
+
+    //generate field element. MISSING null-check
+    field_t* field = zendoo_deserialize_field(mediumTerm);
+    return field;
+#endif
+}
+
+field_t* SidechainTxsCommitmentBuilder::mapCertToField(const uint256& certHash)
+{
+    static_assert(sizeof(uint256) <= SC_FIELD_SAFE_SIZE,
+            "Certificate data to field point mapping not working with current configuration");
+    unsigned char hash[SC_FIELD_SIZE] = {};
+    std::string resHex = std::string(certHash.begin(), certHash.end());
+    memcpy(hash, resHex.append(SC_FIELD_SIZE - resHex.size(), '\0').c_str(), SC_FIELD_SIZE);
+
+    //generate field element. MISSING null-check
+    field_t* field = zendoo_deserialize_field(hash);
+    return field;
+}
+
+uint256 SidechainTxsCommitmentBuilder::mapFieldToHash(const field_t* pField)
+{
+    if (pField == nullptr)
+        throw;
+
+    unsigned char field_bytes[SC_FIELD_SIZE];
+    zendoo_serialize_field(pField, field_bytes);
+
+    uint256 res = Hash(BEGIN(field_bytes), END(field_bytes));
+    return res;
+}
+
+unsigned int SidechainTxsCommitmentBuilder::treeHeightForLeaves(unsigned int numberOfLeaves) const
+{
+    assert(numberOfLeaves > 0);
+
+    return static_cast<unsigned int>(ceil(log2f(static_cast<float>(numberOfLeaves))));
+}
+
+field_t* SidechainTxsCommitmentBuilder::merkleTreeRootOf(std::vector<field_t*>& leaves) const
+{
+    //Notes: Leaves are consumed, i.e. freed and nulled;
+    //       It is guaranteed not to return nullptr.
+
+    unsigned int numberOfLeaves = leaves.size();
```

Comment:
`numberOfLeaves` computation doesn't consider `null` leaves. So in case of null leaves, we will put less entries, than expected in `btrTree`.

Reply:
@i-Alex , thanks for your observation. 
You are correct. Null leaves come out only of error in serialization. Current draft just swallows this error and goes on with trying and building the tree. This will crash upon encountering null leaves.
We will introduce a different handling of serialization, making the add fail in case null leave is returned, so to be able to assert that all leaves submitted to the tree are not null.
Moreover the final tree must have a fixed number of leaves (and a fixed height consequently) which we need yet to determine (I guess give limits on block size and minimal sizes of sidechain related transactions)

Reply:
Just want to add the case with `btrRootField` that is always null for now and it's correct (maybe):
```
std::vector<field_t*> singleSidechainComponents = {ftRootField, btrRootField, certRootField, scIdField};
field_t * sidechainTreeRoot = merkleTreeRootOf(singleSidechainComponents);
```